src/main/java/io/cdap/plugin/couchbase/CouchbaseSourceConfig.java

```diff
@@ -0,0 +1,359 @@
+                             "Specify valid error handling strategy name")
+          .withConfigProperty(CouchbaseConstants.ON_ERROR);
+      }
+    }
+    if (!containsMacro(CouchbaseConstants.SCAN_CONSISTENCY)) {
+      if (Strings.isNullOrEmpty(consistency)) {
+        collector.addFailure("Scan consistency must be specified", null)
+          .withConfigProperty(CouchbaseConstants.SCAN_CONSISTENCY);
+      } else if (Consistency.fromDisplayName(consistency) == null) {
+        collector.addFailure("Invalid scan consistency name", null)
+          .withConfigProperty(CouchbaseConstants.SCAN_CONSISTENCY);
+      }
+    }
+    if (!containsMacro(CouchbaseConstants.QUERY_TIMEOUT)) {
+      if (timeout < 1) {
+        collector.addFailure("Query timeout must be greater than 0", null)
+          .withConfigProperty(CouchbaseConstants.QUERY_TIMEOUT);
+      }
+    }
+    // TODO Couchbase Server 4.5 introduces INFER, a N1QL statement that infers the metadata of documents.
+    // This can be used to infer the Output Schema.
+    if (!containsMacro(CouchbaseConstants.SCHEMA) && Strings.isNullOrEmpty(schema)) {
+      collector.addFailure("Output schema must be specified", null)
+        .withConfigProperty(CouchbaseConstants.SCHEMA);
+    } else if (!containsMacro(CouchbaseConstants.SCHEMA)) {
+      Schema parsedSchema = getParsedSchema();
+      validateSchema(parsedSchema, collector);
+    }
+  }
+
+  private void validateSchema(Schema schema, FailureCollector collector) {
+    List<Schema.Field> fields = schema.getFields();
+    if (fields == null || fields.isEmpty()) {
+      collector.addFailure("Schema must contain at least one field", null)
+        .withConfigProperty(CouchbaseConstants.SCHEMA);
+      return;
+    }
+    for (Schema.Field field : fields) {
+      validateFieldSchema(field.getName(), field.getSchema(), collector);
+    }
+  }
+
+  private void validateFieldSchema(String fieldName, Schema schema, FailureCollector collector) {
+    Schema nonNullableSchema = schema.isNullable() ? schema.getNonNullable() : schema;
+    Schema.Type type = nonNullableSchema.getType();
+    switch (type) {
+      case RECORD:
+        validateSchema(nonNullableSchema, collector);
+        break;
+      case ARRAY:
+        validateArraySchema(fieldName, nonNullableSchema, collector);
+        break;
+      case MAP:
+        validateMapSchema(fieldName, nonNullableSchema, collector);
+        break;
+      default:
+        validateSchemaType(fieldName, nonNullableSchema, collector);
+    }
+  }
+
+  private void validateMapSchema(String fieldName, Schema schema, FailureCollector collector) {
+    Schema keySchema = schema.getMapSchema().getKey();
+    if (keySchema.isNullable() || keySchema.getType() != Schema.Type.STRING) {
+      collector.addFailure("Map keys must be a non-nullable string",
+                           String.format("Change field '%s' to be a non-nullable string", fieldName))
```

Comment:
to be a non-nullable string -> to use a non-nullable string as the map key

Reply:
Changed.